otter/log/spec.py

```diff
@@ -1,10 +1,63 @@
 """
 Format logs based on specification
 """
+import json
+import math
+
+from toolz.dicttoolz import assoc, keyfilter
+from toolz.functoolz import compose, curry
 
 from twisted.python.failure import Failure
 
 
+_json_len = compose(len, json.dumps)
+
+
+def split_execute_convergence(event, max_length=50000):
+    """
+    Try to split execute-convergence event out into multiple events if there
+    are too many CLB nodes, too many servers, or too many steps.
+
+    The problem is mainly the servers, since they take up the most space.
+
+    Experimentally determined that probably logs cut off at around 75k,
+    characters - we're going to limit it to 50k.
+
+    :param dict event: The 'execute-convergence' type event dictionary to split
+    :param int max_length: The maximum length of the entire JSON-formatted
+        dictionary.
+
+    :return: `list` of `tuple` of (`dict`, `str`).  The `dict`s in the tuple
+        represents the spit up event dicts, and the `str` the format string
+        for each.  If the event does not need to be split, the list will only
+        have one tuple.
+    """
+    message = "Executing convergence"
+    if _json_len(event) <= max_length:
+        return [(event, message)]
+
+    events = [(event, message)]
+    large_things = sorted(('servers', 'lb_nodes'),
+                          key=compose(_json_len, event.get),
+                          reverse=True)
+
+    # simplified event which serves as a base for the split out events
+    base_event = keyfilter(
+        lambda k: k not in ('desired', 'servers', 'lb_nodes', 'steps'),
+        event)
+
+    for thing in large_things:
+        split_up_events = split(
+            curry(assoc, base_event, thing), event[thing], max_length,
```

Comment:
Minor nitpick: You can call `assoc(base_event, thing)` directly since all functions in toolz are curried.